src/workerd/api/streams/readable-source-adapter.c++

```diff
@@ -996,135 +1069,265 @@ void ReadableStreamSourceKjAdapter::cancel(kj::Exception reason) {
 }
 
 kj::Promise<void> ReadableStreamSourceKjAdapter::pumpToImpl(
-    WritableSink& output, EndAfterPump end) {
-  static constexpr size_t kMinRead = 8192;
-  static constexpr size_t kMaxRead = 16384;
-  kj::FixedArray<kj::byte, kMaxRead> buffer;
-  // Let's make sure we're in the right state before we start.
-  KJ_DASSERT(state.is<kj::Own<Active>>());
+    kj::Own<Active> active, WritableSink& output, EndAfterPump end) {
+  KJ_DASSERT(active->state.is<Active::Idle>() || active->state.is<Active::Readable>(),
+      "pumpToImpl called when stream is not in an active state.");
+
+  static constexpr size_t DEFAULT_BUFFER_SIZE = 16384;
+  static constexpr size_t MIN_BUFFER_SIZE = 1024;
+  static constexpr size_t MED_BUFFER_SIZE = MIN_BUFFER_SIZE << 6;
+  static constexpr size_t MAX_BUFFER_SIZE = MIN_BUFFER_SIZE << 7;
+  static constexpr size_t MEDIUM_THRESHOLD = 1048576;
+  static_assert(MIN_BUFFER_SIZE < DEFAULT_BUFFER_SIZE);
+  static_assert(DEFAULT_BUFFER_SIZE < MED_BUFFER_SIZE);
+  static_assert(MED_BUFFER_SIZE < MAX_BUFFER_SIZE);
+  static_assert(MAX_BUFFER_SIZE < MEDIUM_THRESHOLD);
+
+  // The minimum read policy to use during the pump. Starts as OPPORTUNISTIC
+  // but will be adjusted based on observed stream behavior.
+  MinReadPolicy minReadPolicy = MinReadPolicy::OPPORTUNISTIC;
+
+  // Our stream may or may not have a known length.
+  size_t bufferSize = DEFAULT_BUFFER_SIZE;
+  kj::Maybe<uint64_t> maybeRemaining = active->stream->tryGetLength(StreamEncoding::IDENTITY);
+  KJ_IF_SOME(length, maybeRemaining) {
+    // Streams that advertise their length SHOULD always tell the truth.
+    // But... on the off change they don't, we'll still try to behave
```

Comment:
s/change/chance/

I wish commit-by-commit reviewing supported in-line suggestion syntax. :(